### PR TITLE
Additional mempool metrics

### LINKED ISSUE #2626: Add metrics for mempool transaction verification and state checks, including failures
### Designs

See below

### From zcashd (use same names)

- [x] `zcash.mempool.size.transactions [gauge]`  
  - Number of transactions in the mempool.
  - Done in https://github.com/ZcashFoundation/zebra/pull/2860
- [x] `zcash.mempool.size.bytes [gauge]`
  - The sum of the serialized transaction sizes, i.e. how much block space they require.
  - Done in https://github.com/ZcashFoundation/zebra/pull/2860

- ~`zcash.mempool.usage.bytes [gauge]`~
  - The dynamic memory usage of the mempool, i.e. how much it is costing the node to track them.
  - TODO: Tricky; will be done later

### Already existing (from tx verifier/downloader) that we should rename

- [x] `mempool.downloaded.transactions.total [counter]` (renamed from `gossip.downloaded.transaction.count`)
  - Number of transactions downloaded for the mempool
- [x] `mempool.pushed.transactions.total [counter]` (renamed from `gossip.pushed.transaction.count`)
  - Number of transactions pushed directly for verification
- [x] `mempool.verified.transactions.total [counter]` (renamed from `gossip.verified.transaction.count`)
  - Number of transactions verified successfully
- [x] `mempool.cancelled.verify.tasks.total [counter]` (renamed from `gossip.cancelled.count`)
  - Number of transactions downloads/verification cancelled
- [x] `mempool.currently.queued.transactions.total [gauge]` (renamed from `gossip.queued.transaction.count`)
  - Number of transactions currently queued for download/verification
  - (also update it after poll_next()!)

### To be created

- [x] `mempool.queued.transactions.total [counter]`
  - Number of transactions ever queued for download/verification
- [x] `mempool.rejected.transaction.ids.total [gauge]`
  - Number of currently rejected transactions
- [x] `mempool.rejected.transaction.ids.bytes [gauge]`
  - The sum of serialized sizes of rejected IDs
  - TODO: we can later replace by memory size; should be pretty similar
- [x] `mempool.failed.verify.tasks.total [counter]`
  - Number of transactions that failed download or verification
- [x] `mempool.gossiped.transactions.total [counter]`
- TODO: number of transactions gossiped to other nodes?

From design brainstorm document:
- Tag each transaction with its version
- Tag each transaction with its received timestamp/age, if that’s easy

### LINKED ISSUE #2627: Track number of inbound/outbound/in mempool transactions

### PR DESCRIPTION
**Pull request by conradoplg**
## Motivation

We need mempool metrics.

### Designs

Which metrics and their names were discussed in https://github.com/ZcashFoundation/zebra/issues/2626

## Solution

- Add new metrics.
- Rename old ones. 
- Expand grafana dashboard with new metrics.
- Remove old "transaction verification" dashboard that was incorporated in the mempool one.

Closes https://github.com/ZcashFoundation/zebra/issues/2626
Closes https://github.com/ZcashFoundation/zebra/issues/2627 (please check. I think it does, since inbound = downloader, outbound = gossiped)

## Review

Required for beta release. Anyone can review.

### Reviewer Checklist

  - [ ] Code implements Specs and Designs
  - [ ] Tests for Expected Behaviour
  - [ ] Tests for Errors

### CODE REVIEW

**Comment by teor2345** [comment]
>     * Remove old "transaction verification" dashboard that was incorporated in the mempool one.

What about transactions that are verified as part of blocks?

**Comment by conradoplg** [comment]
> > ```
> > * Remove old "transaction verification" dashboard that was incorporated in the mempool one.
> > ```
> 
> What about transactions that are verified as part of blocks?

We don't have metrics for those yet. That dashboard only showed mempool transactions

**Review by oxarbitrage — DISMISSED** [review]
I think the dashboard looks great, the code is also ok for me. I am not very familiar with the `PinnedDrop` part and but i guess this is fine. 

Great work.

File: zebrad/src/components/mempool/downloads.rs [hunk 1]
```diff
@@ -312,11 +312,19 @@ where
                         _ => unreachable!("wrong response to transaction request"),
                     };
 
-                    metrics::counter!("gossip.downloaded.transaction.count", 1);
+                    metrics::counter!(
+                        "mempool.downloaded.transactions.total",
+                        1,
+                        "version" => format!("{}",tx.transaction.version()),
```
Comment (teor2345) [inline zebrad/src/components/mempool/downloads.rs:318]: Very nice!

File: zebrad/src/components/mempool/downloads.rs [hunk 2]
```diff
@@ -370,7 +382,11 @@ where
             ?MAX_INBOUND_CONCURRENCY,
             "queued transaction hash for download"
         );
-        metrics::gauge!("gossip.queued.transaction.count", self.pending.len() as _);
+        metrics::gauge!(
+            "mempool.currently.queued.transactions.total",
```
Comment (teor2345) [inline zebrad/src/components/mempool/downloads.rs]: Naming Nitpick: I think we can just use `transactions` as the unit here:
https://prometheus.io/docs/practices/naming/#metric-names

This change also helps distinguish gauges from counters. (Prometheus or grafana might synthesise a `mempool_rejected_transaction_ids_total` counter anyway. I'm not sure exactly when that happens.)


```suggestion
            "mempool.currently.queued.transactions",
```

File: zebrad/src/components/mempool/storage.rs [hunk 3]
```diff
@@ -454,4 +466,21 @@ impl Storage {
         }
         Ok(())
     }
+
+    /// Update metrics related to the rejected lists.
+    ///
+    /// Must be called every time the rejected lists change.
+    fn update_rejected_metrics(&self) {
+        metrics::gauge!(
+            "mempool.rejected.transaction.ids.total",
```
Comment (teor2345) [inline zebrad/src/components/mempool/storage.rs]: Naming Nitpick: I think we can just use `ids` as the unit for this gauge.

```suggestion
            "mempool.rejected.transaction.ids",
```
https://prometheus.io/docs/practices/naming/#metric-names

**Review by teor2345 — COMMENTED** [review]
I have some optional suggestions about counter names.

The `PinnedDrop` looks fine, we're not doing anything unsafe or even risky in the drop impl.

File: zebrad/src/components/mempool/storage.rs (hunk 3 above)
Comment (conradoplg) [reply zebrad/src/components/mempool/storage.rs]: Done in 660fb787

File: zebrad/src/components/mempool/downloads.rs (hunk 2 above)
Comment (conradoplg) [reply zebrad/src/components/mempool/downloads.rs]: Done in 660fb787

**Review by oxarbitrage — APPROVED** [review]

**Pull request merged**